lib/plugins/aws/package/compile/vpc/index.test.js

```diff
@@ -0,0 +1,122 @@
+    let compileInternetGatewayStub;
+    let compileVpcGatewayAttachmentStub;
+    let compileRouteTablesStub;
+    let compileSubnetsStub;
+    let compileNatGatewaysStub;
+    let compileRoutesStub;
+    let compileSubnetRouteTableAssociationsStub;
+    let compileLambdaSecurityGroupStub;
+    let updateFunctionConfigsStub;
+    let compileCloudFormationOutputsStub;
+    let compileSfeOutputsStub;
+
+    beforeEach(() => {
+      compileVpcStub = sinon.stub(awsCompileVpc, 'compileVpc').resolves();
+      compileEipStub = sinon.stub(awsCompileVpc, 'compileEip').resolves();
+      compileInternetGatewayStub = sinon.stub(awsCompileVpc, 'compileInternetGateway').resolves();
+      compileVpcGatewayAttachmentStub = sinon
+        .stub(awsCompileVpc, 'compileVpcGatewayAttachment')
+        .resolves();
+      compileRouteTablesStub = sinon.stub(awsCompileVpc, 'compileRouteTables').resolves();
+      compileSubnetsStub = sinon.stub(awsCompileVpc, 'compileSubnets').resolves();
+      compileNatGatewaysStub = sinon.stub(awsCompileVpc, 'compileNatGateways').resolves();
+      compileRoutesStub = sinon.stub(awsCompileVpc, 'compileRoutes').resolves();
+      compileSubnetRouteTableAssociationsStub = sinon
+        .stub(awsCompileVpc, 'compileSubnetRouteTableAssociations')
+        .resolves();
+      compileLambdaSecurityGroupStub = sinon
+        .stub(awsCompileVpc, 'compileLambdaSecurityGroup')
+        .resolves();
+      updateFunctionConfigsStub = sinon.stub(awsCompileVpc, 'updateFunctionConfigs').resolves();
+      compileCloudFormationOutputsStub = sinon
+        .stub(awsCompileVpc, 'compileCloudFormationOutputs')
+        .resolves();
+      compileSfeOutputsStub = sinon.stub(awsCompileVpc, 'compileSfeOutputs').resolves();
+    });
+
+    afterEach(() => {
+      awsCompileVpc.compileVpc.restore();
+      awsCompileVpc.compileEip.restore();
+      awsCompileVpc.compileInternetGateway.restore();
+      awsCompileVpc.compileVpcGatewayAttachment.restore();
+      awsCompileVpc.compileRouteTables.restore();
+      awsCompileVpc.compileSubnets.restore();
+      awsCompileVpc.compileNatGateways.restore();
+      awsCompileVpc.compileRoutes.restore();
+      awsCompileVpc.compileSubnetRouteTableAssociations.restore();
+      awsCompileVpc.compileLambdaSecurityGroup.restore();
+      awsCompileVpc.updateFunctionConfigs.restore();
+      awsCompileVpc.compileCloudFormationOutputs.restore();
+      awsCompileVpc.compileSfeOutputs.restore();
+    });
+
+    it('should have hooks', () => expect(awsCompileVpc.hooks).to.be.not.empty);
+
+    it('should set the provider variable to be an instanceof AwsProvider', () =>
+      expect(awsCompileVpc.provider).to.be.instanceof(AwsProvider));
+
+    describe('"package:compileVpc" promise chain', () => {
+      it('should run the promise chain in order', () => {
+        return awsCompileVpc.hooks['package:compileVpc']().then(() => {
+          expect(compileVpcStub.calledOnce).to.be.equal(true);
```

Comment:
To be honest, I'm not convinced whether such tests are really valid

As what is tested here, is not the outcome that really counts: the expected VPC configuration that lands in CF template, but instead we test, whether given internal methods were called and how many times.
It seems that what we're really guarding here, is that this specific implementation is written line by line written as it's currently, and it's not the thing we should really care about I believe

It raises maintenance issues, as whenever we will refactor the implementation, so e.g.  it's leaner (without changing the outcome), those tests will have to be refactored as well - that makes future improvements to the framework more cumbersome to do.

Ideally tests should guard the outcome (desired effect), and not internals of implementation. They should allow to do painless refactors, where we can overhaul the implementation to something very different, and via test run confirm that all works as is it worked before.

Imo right approach is to treat the internals of a plugin as black box. Test the expected result, and via coverage confirm that we covered all cases envisioned in implementation.

I would prepare just one test file for VPC plugin, and would rely on `@serverless/test/run-serverless` util, which ensures you run complete instance, so eventual side effects introduced by other plugins are also in play.

Currently it'll require to create you a dummy service folder (with simplest possible `serverless.yml` that has `vpc: true`) (as side note, we can improve that util so it takes config via argument (instead of `cwd`)).

Having that you may prepare a `serverless` instance ready for testing as:

```javascript
runServerless({
  cwd: vpcServicePath,
  pluginPathsWhitelist: [vpcPluginPath],
  hookNamesWhitelist: ['package:compileVpc'],
})).then(serverless => {
  const cfTemplate = serverless.service.provider.compiledCloudFormationTemplate;
  // Confirm that cfTemplate has VPC configuration we're after
});
```

What do you think about that?

Reply:
Totally agree with that. I guess the root cause for such tests were the chasing of test coverage increases in the past.

So the idea is that I completely remove the unit tests for this plugin? I'm not sure about that because for large implementations (like this) it can be really tough to track down the actual root cause for the test failure.

But yes, let me prepare such test with our new `runServerless` capability (really like the API 👍).

Reply:
> So the idea is that I completely remove the unit tests for this plugin?  

Yes, I would ditch all tests that check whether we called given function and how many times 
and just run tests that confirm on generated template.

> I'm not sure about that because for large implementations (like this) it can be really tough to track down the actual root cause for the test failure.

I wonder whether that can be a real issue. Technically implementation should be quite self explanatory, so if we see that that there's an issue in generated resource X or that it's missing, it shouldn't be a big trouble to track down the internal function that's responsible for that.

At least testing that depends on specificity of internals, puts on us hand-cuffs, that whenever we shuffle internals, we also should update tests, and that doesn't say good of tests imo.

On unit tests level, I would only leave tests that confirm some non trivial algorithms (e.g. fahrenheit o celsius conversion, deciding which year is leap year). Still in many cases such functions are rather meant to be excluded into utils, and over there such unit tests naturally make big sense.

So I guess good approach would be:
- Test outcome of a plugin as if it's a black box (tests should not check on it's implementation internals, how many and what functions are created internally and which order invoked)
- Exclude any internal non trivial logic units into utils (if it's questionable to expose them globally we may create `utils.js` in plugin directory), which should be accompanied with unit tests that test this util individually. (Still I don't think there's a case of such utils in context of this plugin ?)

Reply:
I agree with the sentiment to focus on integration test rather than starting with unit tests. However very often there are nuanced edge cases which would require a lot of different integration test setups. Such edge cases are easier tested with unit tests and then the integration test is used to only cover the "happy" / simple path.

I'd suggest to default to integration tests only but if there are some non-trivial edge cases we should mix in unit tests for such scenarios.

Reply:
Sure, if for some reason just via natural plugin evaluation, it's hard to test some scenarios, then indeed we probably should explore alternatives, and maybe introduce unit tests for some isolated part.

Still even in such case, I think it'll be good to ensure that tests do not depend too much on  structure of internals.

Do we have such cases here?

Reply:
> Still even in such case, I think it'll be good to ensure that tests do not depend too much on structure of internals.

Makes sense 👍

> Do we have such cases here?

Not that I'm aware of 🤔

Reply:
Just removed the unit tests and replaced them with an integration test.